### ISSUE TITLE
Browser WebExtension with Native Messaging as alternative to Electron

### ISSUE THREAD

**Issue by stoically** [vector-im/riot-web#14523]
**Is your suggestion related to a problem? Please describe.**

Personally I prefer running web apps in my browser of choice, which is not possible with apps that are powered by Electron. Running in the browser also has several benefits like
- let apps share the system resources of one active browser
- allows grouping apps into pinned tabs or windows

Using Element Web as is however comes with a set of limitations, namely it doesn't provide search in encrypted rooms (seshat) or storing of pickle keys into OS secure storage (keytar).

**Describe the solution you'd like**

To get those features that rely on native code into the browser world, something that can be done is using a [WebExtension](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions) that talks via [Native messaging](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging) with a executable on the host through stdio.

When trying to use a functionality that normally is restricted to Element Desktop, like when searching in an encrypted room, Element Web could offer instructions to install the WebExtension along with the native messaging host binary.

**Describe alternatives you've considered**

Not aware of alternatives yet.

**Comment by steef435** [comment]
If this would replace the electron app, it would be rather unhandy with three separate dependencies (browser, extension, native binary), all with their own update mechanism.

**Comment by t3chguy** [comment]
It most definitely would not replace it.

**Comment by stoically** [comment]
Yeah, the idea would be just to have an alternative, not a replacement.

Keeping the browser up to date should be done regardless, the extension can auto-update itself, and yeah, the native binary would be a bit more tricky - though, it could be made self-updating as well.

**Comment by jryans** [comment]
@stoically Shouldn't you mention your own extensions which implement effectively this...? At the very least, you've already prototyped the idea.

**Comment by stoically** [comment]
Didn't want to make the impression of using this issue to promote my extensions. Would be nice to see it upstream and hence thought it might make sense to have a place to discuss this as a concept.

**Comment by t3chguy** [comment]
# @stoically has great Web Extensions which solve this issue such as https://github.com/stoically/radical-native

-----

No worry, I'll do it for you :P

**Comment by jryans** [comment]
Aha okay, sorry then. I was just a bit surprised to see no mention of it, as surely it was be at least a good reference for anyone thinking about this idea!
